shapely/set_operations.py

```diff
@@ -164,8 +166,19 @@ def intersection_all(geometries, axis=None, **kwargs):
     <LINESTRING (1 1, 2 2)>
     >>> intersection_all([[line1, line2, None]], axis=1).tolist()
     [<LINESTRING (1 1, 2 2)>]
+    >>> intersection_all([line1, None])
+    <GEOMETRYCOLLECTION EMPTY>
```

Comment:
This rule of "at least two non-None geometries", that gives an inconsistency with having only one geometry?

```
In [31]: shapely.intersection_all([line1])
Out[31]: <LINESTRING (0 0, 2 2)>
```

(and I know this is already the case right now, but since we are discussing all those corner cases here ... ;))

Also, I would expect the following two to result in the same (the below is using main, not this PR):

```
In [32]: shapely.intersection_all([line1, None])   # returns None

In [33]: shapely.intersection_all([line1, None, None])
Out[33]: <LINESTRING (0 0, 2 2)>
```

It seems a bit strange that a single None or two or more Nones are treated differently here. 
With this PR, those both will give an empty geometry now? That at least makes it consistent, but I would personally still expect it to be the same as the first case (`shapely.intersection_all([line1])`), if we consider the logic to be "Nones are skipped".

Reply:
Testing with your PR, they do _not_ both give an empty geometry:

```
In [3]: shapely.intersection_all([line1, None, None])
Out[3]: <LINESTRING (0 0, 2 2)>

In [4]: shapely.intersection_all([line1, None])
Out[4]: <GEOMETRYCOLLECTION EMPTY>

In [5]: shapely.intersection_all([line1])
Out[5]: <LINESTRING (0 0, 2 2)>

```

_If_ we would want to make this consistent (all return the linestring?), I suppose that needs a change in the C code?

Reply:
The problem is that trying to reduce inputs of length 2 no longer fall into the `YY_Y_func_reduce` function in C, which specifically handles `None` and ignores cases.  They instead fall through to the regular `YY_Y_func` function in C, where we explicitly return `None` if either input is `None`.  Given that this is based on whatever is the inner axis of the operation, it is tough to special case this by filtering `None` out in advance.

Perhaps we also need to revisit if `intersection(Point(0, 0), None)` should return the non-`None` input, or always return `None` (as it does now).  This affects intersection, union, difference, symmetric_difference, and shared_paths.

Reply:
I think that the bug in #1513 necessitated the `dimensions[0] > 1` check here. Then at least `intersection_all([geom])` does not segfault. Now that the bug is fixed I see no reason in keeping the `dimensions[0] > 1` check.